app/src/main/java/fr/free/nrw/commons/category/CategorizationFragment.java

```diff
@@ -320,8 +321,8 @@ protected void onPostExecute(ArrayList<String> result) {
                 latch.countDown();
             }
         };
-        prefixUpdaterSub.executeOnExecutor(AsyncTask.THREAD_POOL_EXECUTOR);
-        methodAUpdaterSub.executeOnExecutor(AsyncTask.THREAD_POOL_EXECUTOR);
+        prefixUpdaterSub.executeOnExecutor(Executors.newSingleThreadExecutor());
+        methodAUpdaterSub.executeOnExecutor(Executors.newSingleThreadExecutor());
     }
```

Comment:
Or the new idea comes to my mind: using limited number of threads, when one of them hung, we can detect by limit exceed.